File: net/JetBrains.SignatureVerifier/ReadUtils.cs
using System.IO;

namespace JetBrains.SignatureVerifier
{
  static class ReadUtils
  {
    /// <summary>
    /// Set the stream position to start
    /// </summary>
    /// <param name="stream">Stream</param>
    /// <returns>Fluent</returns>
    internal static Stream Rewind(this Stream stream)
    {
      stream.Seek(0, SeekOrigin.Begin);
      return stream;
    }

    /// <summary>
    /// Read all data from the stream
    /// </summary>
    /// <param name="stream">Stream</param>
    internal static byte[] ReadAll(this Stream stream)
    {
      stream.Rewind();
      using var ms = new MemoryStream();
      stream.CopyTo(ms);
      return ms.ToArray();
    }

    /// <summary>
    /// Read all data from the current position of the stream
    /// </summary>
    /// <param name="stream">Stream</param>
    internal static byte[] ReadToEnd(this Stream stream)
    {
      using var ms = new MemoryStream();
      stream.CopyTo(ms);
      return ms.ToArray();
    }

    internal static uint ReadUInt32Le(BinaryReader reader, bool isBe)
    {
      var value = reader.ReadUInt32();
      return isBe
        ? SwapBytes(value)
        : value;
    }

    internal static ulong ReadUInt64Le(BinaryReader reader, bool isBe)
    {
      var value = reader.ReadUInt64();
      return isBe
        ? SwapBytes(value)
        : value;
    }

    private static ushort SwapBytes(ushort val)
    {
      return (ushort)((val << 8) | (val >> 8));
    }

    private static uint SwapBytes(uint val)
    {
      return (uint)SwapBytes((ushort)val) << 16 |
             (uint)SwapBytes((ushort)(val >> 16));
    }

    private static ulong SwapBytes(ulong val)
    {
      return (ulong)SwapBytes((uint)val) << 32 |
             (ulong)SwapBytes((uint)(val >> 32));
    }
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/ReadUtils.kt
﻿package com.jetbrains.signatureverifier

import java.nio.ByteBuffer
import java.nio.channels.SeekableByteChannel

fun SeekableByteChannel.Rewind(): SeekableByteChannel {
  this.position(0)
  return this
}

fun SeekableByteChannel.Seek(position: Long, origin: SeekOrigin) {
  when (origin) {
    SeekOrigin.Begin -> this.position(position)
    SeekOrigin.Current -> this.position(this.position() + position)
    SeekOrigin.End -> this.position(this.size() - 1)
  }
}

fun SeekableByteChannel.ReadAll(): ByteArray {
  this.Rewind()
  val buf = ByteBuffer.wrap(ByteArray(this.size().toInt()))
  this.read(buf)
  return buf.array()
}

fun SeekableByteChannel.ReadToEnd(): ByteArray {
  val pos = this.position()
  val size = this.size()
  if (pos + 1 == size)
    return ByteArray(0)
  val buf = ByteBuffer.wrap(ByteArray((size - pos).toInt()))
  this.read(buf)
  return buf.array()
}

fun BinaryReader.ReadUInt32Le(isBe: Boolean): UInt {
  val value = this.ReadUInt32()
  return if (isBe) SwapBytes(value) else value
}

fun BinaryReader.ReadUInt64Le(isBe: Boolean): ULong {
  val value = this.ReadUInt64()
  return if (isBe) SwapBytes(value) else value
}

private fun SwapBytes(v: UShort): UShort = (((v.toInt() and 0xFF) shl 8) or ((v.toInt() ushr 8) and 0xFF)).toUShort()

private fun SwapBytes(v: UInt): UInt =
  (SwapBytes(v.toUShort()).toUInt() shl 16) or (SwapBytes((v shr 16).toUShort()).toUInt())

private fun SwapBytes(v: ULong): ULong =
  (SwapBytes(v.toUInt()) shl 32).toULong() or (SwapBytes(v.toUInt()) shr 32).toULong()
